channeldb/migration13/migration.go

```diff
@@ -0,0 +1,202 @@
+package migration13
+
+import (
+	"encoding/binary"
+	"fmt"
+
+	"github.com/coreos/bbolt"
+)
+
+var (
+	paymentsRootBucket = []byte("payments-root-bucket")
+
+	// paymentCreationInfoKey is a key used in the payment's sub-bucket to
+	// store the creation info of the payment.
+	paymentCreationInfoKey = []byte("payment-creation-info")
+
+	// paymentFailInfoKey is a key used in the payment's sub-bucket to
+	// store information about the reason a payment failed.
+	paymentFailInfoKey = []byte("payment-fail-info")
+
+	// paymentAttemptInfoKey is a key used in the payment's sub-bucket to
+	// store the info about the latest attempt that was done for the
+	// payment in question.
+	paymentAttemptInfoKey = []byte("payment-attempt-info")
+
+	// paymentSettleInfoKey is a key used in the payment's sub-bucket to
+	// store the settle info of the payment.
+	paymentSettleInfoKey = []byte("payment-settle-info")
+
+	// paymentHtlcsBucket is a bucket where we'll store the information
+	// about the HTLCs that were attempted for a payment.
+	paymentHtlcsBucket = []byte("payment-htlcs-bucket")
+
+	// htlcAttemptInfoKey is a key used in a HTLC's sub-bucket to store the
+	// info about the attempt that was done for the HTLC in question.
+	htlcAttemptInfoKey = []byte("htlc-attempt-info")
+
+	// htlcSettleInfoKey is a key used in a HTLC's sub-bucket to store the
+	// settle info, if any.
+	htlcSettleInfoKey = []byte("htlc-settle-info")
+
+	// htlcFailInfoKey is a key used in a HTLC's sub-bucket to store
+	// failure information, if any.
+	htlcFailInfoKey = []byte("htlc-fail-info")
+
+	byteOrder = binary.BigEndian
+)
+
+// MigrateMPP migrates the payments to a new structure that accommodates for mpp
+// payments.
+func MigrateMPP(tx *bbolt.Tx) error {
+	log.Infof("Migrating payments to mpp structure")
+
+	// Iterate over all payments and store their indexing keys. This is
+	// needed, because no modifications are allowed inside a Bucket.ForEach
+	// loop.
+	paymentsBucket := tx.Bucket(paymentsRootBucket)
+	if paymentsBucket == nil {
+		return nil
+	}
+
+	var paymentKeys [][]byte
+	err := paymentsBucket.ForEach(func(k, v []byte) error {
+		paymentKeys = append(paymentKeys, k)
+		return nil
+	})
+	if err != nil {
+		return err
+	}
+
+	// With all keys retrieved, start the migration.
+	for _, k := range paymentKeys {
+		bucket := paymentsBucket.Bucket(k)
+
+		// We only expect sub-buckets to be found in
+		// this top-level bucket.
+		if bucket == nil {
+			return fmt.Errorf("non bucket element in " +
+				"payments bucket")
+		}
+
+		// Fetch old format creation info.
+		creationInfo := bucket.Get(paymentCreationInfoKey)
+		if creationInfo == nil {
+			return fmt.Errorf("creation info not found")
+		}
+
+		// Make a copy because bbolt doesn't allow this value to be
+		// changed in-place.
+		newCreationInfo := make([]byte, len(creationInfo))
+		copy(newCreationInfo, creationInfo)
+
+		// Convert to nano seconds.
+		timeBytes := newCreationInfo[32+8 : 32+8+8]
```

Comment:
Non-blocking but this section could be made more readable by getting ride of the magic numbers and using types from the `time` package.